packages/swing-store/src/repair-metadata.js

```diff
@@ -0,0 +1,82 @@
+import { Fail, q } from '@agoric/assert';
+import { makeSwingStore } from './swingStore.js';
+import { assertComplete } from './importer.js';
+
+/**
+ * Given a pre-existing swingstore and a SwingStoreExporter, read in
+ * all the metadata from the exporter and use it to regenerate any
+ * missing metadata records. This can be used to fix the damage caused
+ * by #8025.
+ *
+ * The repair method will call `exporter.getExportData` and examine
+ * all entries to do one of three things:
+ *
+ * 1: kvStore records are ignored (not metadata)
+ * 2: bundle/snapshot/transcript records whose keys already exist will
+ *    be compared against the existing data, and an error thrown if
+ *    they do not match
+ * 3: new bundle/snapshot/transcript records will be silently added to
+ *    the swingstore
+ *
+ * It will not call `exporter.getArtifactNames` or `getArtifacts`.
+ *
+ * At the end of the process, the DB changes will be committed. The
+ * function does not return a SwingStore: most callers will want to
+ * call `openSwingStore` shortly after this repair process.
```

Comment:
Please return the swing-store and not commit. The importer does its own consistency check on the `blockHeight` saved in the host section before commiting.

Reply:
Thinking about this more, I can live with an implicit commit, as long as only non-current metadata is checked for consistency on repair.

Reply:
I'll change this to whatever you think is best.

One downside of not doing the commit locally is that it becomes the caller's responsibility, and doing `await repairMetadata(..)` instead of `const { hostStorage } = await repairMetadata(..); await hostStorage.commit()` is a great way to silently discard all the changes (which, since it's merely a repair and not a populate-everything, could easily go unnoticed).

I'm not sure I want to restrict `assertComplete()` to only checking non-current metadata.. in fact the current code only checks *current* data (it asserts that all bundles are populated, all current transcript spans are fully populated, and all current snapshots are populated, so it's looking at more than just metadata). So if the checks are likely to interfere with the cosmic-swingset use case, we should change the behavior.

(I can't think of any way that repair-metadata.js would have opinions on the `blockHeight` field, or the semantics of anything else in kvStore)

Reply:
As discussed, we'll handle the hangover case by checking the block height matches at the cosmos / cosmic-swingset layer. For that however it'd be easier to return a swing-store before commit, and defer commiting to cosmic-swingset

Reply:
I think I understand the hangover problem, but as you say our defense against corruption is to check the block height before beginning the export-data IAVL migration.

If I understand it correctly, it's not that the caller of `repairMetadata()` needs the swingstore code to refrain from doing a commit, but rather it's that the caller needs to check the blockHeight before moving on to the next step, and to do that, they need a `SwingStore` instance, and it would be more convenient for that code to do:
```js
const swingStore = await repairMetadata(dirPath, exporter);
await swingStore.hostStorage.commit();
const ssHeight = swingStore.hostStorage.kvStore.get('blockHeight');
compare(ssHeight, iavlHeight);
const exporter = makeSwingStoreExporter(dirPath);
// delete old IAVL data, use exporter.getExportData() to repopulate
```

than to do:
```js
await repairMetadata(dirPath, exporter);
const swingStore = openSwingStore(dirPath);
const ssHeight = swingStore.hostStorage.kvStore.get('blockHeight');
compare(ssHeight, iavlHeight);
const exporter = makeSwingStoreExporter(dirPath);
// delete old IAVL data, use exporter.getExportData() to repopulate
```

Writing it out that way, I think having `repairMetadata()` do the commit internally looks cleaner. And requiring the caller to remember to `commit()` feels like a bug farm.

What's your current thinking here?

Reply:
The example above is wrong, I'd basically have done:
```js
const swingStore = await repairMetadata(dirPath, exporter);
const ssHeight = swingStore.hostStorage.kvStore.get('blockHeight');
compare(ssHeight, iavlHeight);
await swingStore.hostStorage.commit();
await swingStore.hostStorage.close();
// return to golang, which will initiate the other steps independently, 
// like delete old IAVL data, and start a new export to repopulate
```

Right now I have to explicitly open the DB instead to do the check, and decided to do it first:
https://github.com/Agoric/agoric-sdk/pull/8143/commits/de0c3e6fd0eeded3797a958be9662a6ab09ebef2

So roughly:
```js
const swingStore = openSwingStore(dirPath);
const ssHeight = swingStore.hostStorage.kvStore.get('blockHeight');
compare(ssHeight, iavlHeight);
await swingStore.hostStorage.close();
await repairMetadata(dirPath, exporter);
// return to golang, which will initiate the other steps independently,
// like delete old IAVL data, and start a new export to repopulate
```

Thinking about this more, the most logical thing to do would be for `repairMetadata` to be a method on an already opened swing-store, so the check can be done first, then kick off the repair, and like any operation mutating an open DB, still require an explicit `commit()`, like so:

```js
const swingStore = openSwingStore(dirPath);
const ssHeight = swingStore.hostStorage.kvStore.get('blockHeight');
compare(ssHeight, iavlHeight);
await swingStore.repairMetadata(exporter);
await swingStore.hostStorage.commit();
await swingStore.hostStorage.close();
// return to golang, which will initiate the other steps independently,
// like delete old IAVL data, and start a new export to repopulate
```

Reply:
I don't think those `hostStorage.close()` calls are correct: that does `db = null` internally, which means the entire `swingStore` instance is now dead; `close()` is meant to be the very last thing you do before dropping the `swingStore` entirely.

But apart from that, your last example seems good to me: open a swingstore, assert that the height matches the IAVL's value (otherwise the repair might be populating the wrong metadata records, if a snapshot or span was created between the two heights), do the repair, commit, then *don't* `close()`, and continue to use the same `swingStore` instance for everything else the host wants to do.

I'll add `repairMetadata` as a `hostStorage` method, next to the `commit` that the caller must use afterwards (and since `setExportCallback` is sort of related to metadata anyways).